std-bits/base/src/main/java/org/enso/base/parser/NumberWithSeparators.java

```diff
@@ -0,0 +1,393 @@
+    var format =
+        switch (separator) {
+          case '.' -> DOT_COMMA;
+          case ',' -> separatorCount == 1 ? COMMA_UNKNOWN : COMMA_DOT;
+          case ' ' -> (decimal == Constants.UNKNOWN
+              ? SPACE_UNKNOWN
+              : (decimal == '.' ? SPACE_DOT : SPACE_COMMA));
+          case '\'' -> (decimal == Constants.UNKNOWN
+              ? SWISS_UNKNOWN
+              : (decimal == '.' ? SWISS_DOT : SWISS_COMMA));
+          default -> null;
+        };
+    if (format == null) {
+      return new NumberParseFailure("No matching number format.");
+    }
+
+    var result = format.parseFixedInteger(value, idx, endIdx, separator);
+    return (result instanceof NumberParseFailure)
+        ? result
+        : new NumberParseResultWithFormat(format, result);
+  }
+
+  /** Given a known double format, parse the sequence. */
+  private NumberParseResult parseFixed(
+      CharSequence value, int idx, int endIdx, char firstSeparator, char secondSeparator) {
+    if (this == DOT_UNKNOWN || this == UNKNOWN_DOT) {
+      assert firstSeparator == '.' && secondSeparator == Constants.NONE;
+      return NO_DOT.parseFixed(value, idx, endIdx, Constants.NONE, '.');
+    } else if (this == COMMA_UNKNOWN) {
+      assert firstSeparator == ',' && secondSeparator == Constants.NONE;
+      return COMMA_DOT.parseFixed(value, idx, endIdx, ',', '.');
+    } else if (this == UNKNOWN_COMMA) {
+      assert firstSeparator == ',' && secondSeparator == Constants.NONE;
+      return NO_COMMA.parseFixed(value, idx, endIdx, Constants.NONE, ',');
+    }
+
+    assert thousands != Constants.UNKNOWN && decimal != Constants.UNKNOWN;
+
+    // Validate Separators.
+    if (firstSeparator != Constants.NONE) {
+      if ((secondSeparator == Constants.NONE
+              && firstSeparator != thousands
+              && firstSeparator != decimal)
+          || (secondSeparator != Constants.NONE
+              && (firstSeparator != thousands || secondSeparator != decimal))) {
+        return new NumberParseFailure("Invalid separator.");
+      }
+    }
+
+    // Strip out the separators.
+    int origEndIdx = endIdx;
+    if (thousands != Constants.NONE || decimal != '.') {
+      value = Separators.strip(value, idx, endIdx, thousands, decimal);
+      if (value == null) {
+        return new NumberParseFailure("Invalid number.");
+      }
+      idx = 0;
+      endIdx = value.length();
+    }
+
+    try {
+      double number = Double.parseDouble(value.subSequence(idx, endIdx).toString());
+      return new NumberParseResultWithIndex(origEndIdx, new NumberParseDouble(number, ""));
+    } catch (NumberFormatException e) {
+      return new NumberParseFailure("Invalid number.");
+    }
+  }
+
+  /** Given a unknown format, parse the sequence. */
+  private NumberParseResult parseUnknown(
+      CharSequence value,
+      int idx,
+      int endIdx,
+      char firstSeparator,
+      char secondSeparator,
+      int separatorCount,
+      int lastSeparator) {
```

Comment:
```suggestion
      int lastSeparatorIdx) {
```

`firstSeparator` is a separator character, but `lastSeparator` is a position. It is not clear from the names and the types are not immediately visible when reading the code below. I'd make this variable have `Idx` suffix to make it clear what it is.